### ISSUE TITLE
warn when redundant temporary lists are used when generators would work

### ISSUE THREAD

**Issue by vapier** [PyCQA/pylint#2447]
### Is your feature request related to a problem? Please describe

people often use temporary lists in contexts where generators are acceptable.  for example:
```python
# Replace ... with any generator expression like 'str(x) for x in some_list'.
some_list.extend([...])
some_set.add([...])
''.join([...])
set([...])
all([...])
any([...])
```

this will create a temporary list which is then passed to the respective functions which will consume the list.  if a generator was passed in directly, it would have been cheaper at runtime (no need to create the temporary list, so the content would have been ingested directly), and for some funcs (like `all` or `any`), it prevents the function from returning early.

### Describe the solution you'd like

whenever pylint sees temporary lists in constructs it knows a generator will work, throw a warning telling people to pass a generator in directly.

**Comment by r3m0t** [comment]
The behaviour is slightly different:

```python
def f(x):
    if x == 4: raise IOError
    return x

xs = []
try:
    xs.extend([f(x) for x in range(10)])
except IOError:
    pass

assert xs == []


xs = []
try:
    xs.extend(f(x) for x in range(10))
except IOError:
    pass
assert xs == list(range(4))
```

**Comment by PCManticore** [comment]
Ah, that's a great point @r3m0t  I think though we should still do it for `any` and `all`.

**Comment by vapier** [comment]
i'd contend that this change in behavior doesn't mean we shouldn't still warn about these scenarios.  if people want to make sure the list is fully constructed before extending, they can express that with an explicit `list(...)` wrapper around the generator.

**Comment by Pierre-Sassoulas** [comment]
This was implemented independently from a duplicate https://github.com/PyCQA/pylint/issues/3165 in https://github.com/PyCQA/pylint/pull/3309 in pylint 2.7.0
